### PR TITLE
[segwit] Serialize transactions sent without using inv

### PR DESCRIPTION
**Pull request by xenog**
Extra SegWit-related improvements.

### CODE REVIEW

**Comment by xenog** [comment]
Test is failing on bloom filter membership. Since bloom filters are probabilistic data structures and false positives are possible, this test may end up passing if rescheduled.

File: .gitignore [hunk 1]
```diff
@@ -6,3 +6,4 @@ target
 *.iml
 *.chain
 *.spvchain
+*.wallet
```
Comment (schildbach) [inline .gitignore]: This change seems unrelated.

**Comment by schildbach** [comment]
Thanks for implementing Bech32, it's great to see the puzzle being completed piece by piece!

I think the SegwitAddress class should somehow fit into the VersionChecksummedBytes hierarchy. My idea would be to separate the "legacy base58"-aspects of Address into a subclass LegacyAddress and then make SegwitAddress a subclass of Address. The reason is we will have many cases where we want to use different types of addresses in a list, or want to use the same UI elements, etc.

Also I think perhaps the Bech32 + SegwitAddress changes could go into a separate PR so we don't mix too many different changes.

**Comment by xenog** [comment]
I agree that it would be nice to have to have them in the same hierarchy and behaving in a similar manner. I am not sure how to go about it since it seems that `VersionChecksummedBytes` is very tightly-coupled with base58 semantics. If you factor out the base58 stuff, I'll do the necessary changes to `SegWitAddress`.

File: core/src/main/java/org/bitcoinj/core/BitcoinSerializer.java [hunk 2]
```diff
@@ -205,7 +205,7 @@ private Message makeMessage(String command, int length, byte[] payloadBytes, byt
         } else if (command.equals("getheaders")) {
             message = new GetHeadersMessage(params, payloadBytes);
         } else if (command.equals("tx")) {
-            message = makeTransaction(payloadBytes, 0, length, hash);
```
Comment (schildbach) [inline core/src/main/java/org/bitcoinj/core/BitcoinSerializer.java]: I think this will be fixed once #1438 is merged. Would you like to review that PR?

File: core/src/main/java/org/bitcoinj/core/Transaction.java [hunk 3]
```diff
@@ -276,6 +276,25 @@ public Transaction(
         super(params, payload, 0, parent, setSerializer, length);
     }
 
+    public Transaction disableWitnessSerialization() {
```
Comment (schildbach) [inline core/src/main/java/org/bitcoinj/core/Transaction.java]: I think this method is missing a JavaDoc.

**Comment by schildbach** [comment]
Can you rebase this on the segwit branch? I lost track on what is left of this PR, and the title/description doesn't really tell.

**Comment by xenog** [comment]
I rebased this branch as requested.

File: core/src/main/java/org/bitcoinj/core/Peer.java [hunk 4]
```diff
@@ -1469,6 +1469,7 @@ private void blockChainDownloadLocked(Sha256Hash toHash) {
                     this, toHash, chainHead.getHeader().getHashAsString());
         StoredBlock cursor = chainHead;
         for (int i = 100; cursor != null && i > 0; i--) {
+            //
```
Comment (schildbach) [inline core/src/main/java/org/bitcoinj/core/Peer.java]: This change seems unrelated.

File: core/src/test/java/org/bitcoinj/core/PeerTest.java [hunk 5]
```diff
@@ -94,7 +94,7 @@ public void tearDown() throws Exception {
     }
 
     private void connect() throws Exception {
-        connectWithVersion(70001, VersionMessage.NODE_NETWORK);
+        connectWithVersion(70012, VersionMessage.NODE_NETWORK | VersionMessage.NODE_WITNESS);
```
Comment (schildbach) [inline core/src/test/java/org/bitcoinj/core/PeerTest.java]: This change changes all the tests. I wonder if we should configure this class such that tests are run twice, once with witness/WitnessBlock and once without?

File: core/src/test/java/org/bitcoinj/core/PeerTest.java [hunk 6]
```diff
@@ -373,7 +373,7 @@ public synchronized void onBlocksDownloaded(Peer p, Block block, @Nullable Filte
         List<InventoryItem> items = getdata.getItems();
         assertEquals(1, items.size());
         assertEquals(b2.getHash(), items.get(0).hash);
-        assertEquals(InventoryItem.Type.Block, items.get(0).type);
+        assertEquals(InventoryItem.Type.WitnessBlock, items.get(0).type);
     }
```
Comment (schildbach) [inline core/src/test/java/org/bitcoinj/core/PeerTest.java]: Same as above.

File: core/src/main/java/org/bitcoinj/core/BitcoinSerializer.java [hunk 7]
```diff
@@ -231,7 +231,7 @@ private Message makeMessage(String command, int length, byte[] payloadBytes, byt
         } else if (command.equals("getutxos")) {
             return new GetUTXOsMessage(params, payloadBytes);
         } else if (command.equals("sendheaders")) {
-            return new SendHeadersMessage(params, payloadBytes);
+            return new SendHeadersMessage();
```
Comment (schildbach) [inline core/src/main/java/org/bitcoinj/core/BitcoinSerializer.java]: This change seems unrelated.

File: core/src/test/java/org/bitcoinj/core/PeerTest.java (hunk 5 above)
Comment (xenog) [reply core/src/test/java/org/bitcoinj/core/PeerTest.java]: Is the node itself ever expected to have the `NODE_WITNESS` flag disabled after the fork?

File: core/src/main/java/org/bitcoinj/core/Peer.java (hunk 4 above)
Comment (xenog) [reply core/src/main/java/org/bitcoinj/core/Peer.java]: One that fell through the cracks.

File: core/src/main/java/org/bitcoinj/core/BitcoinSerializer.java (hunk 7 above)
Comment (xenog) [reply core/src/main/java/org/bitcoinj/core/BitcoinSerializer.java]: Another one that fell through the cracks.

**Comment by xenog** [comment]
I removed unrelated changes, but I think we should not disable the `NODE_WITNESS` flag anymore in the P2P code. SegWit is active in the main network, and it would be pointless to disable this flag until a protocol upgrade is done where the flag becomes implicit, or if we want to maintain Bitcoin Cash compatibility in the same code branch, in which case I wolud recommend adding Bitcoin Cash tests separately anyway.

File: core/src/test/java/org/bitcoinj/core/PeerTest.java (hunk 5 above)
Comment (schildbach) [reply core/src/test/java/org/bitcoinj/core/PeerTest.java]: I can't answer this question for sure, but I assume yes. Bitcoinj is used in networks that don't have segwit (yet). Also even in the Bitcoin P2P network peer versions <70012 will exist for quite some years.

One also could argue that simply because we have the code/ability to do so we should continue to have tests for it. It seems to me that we now have quite some `if (segwit) { … } else { … }` in code, and it feels wrong to me to drop tests we already have for the else branches.

**Comment by schildbach** [comment]
Ok, don't worry. I merged. Thanks!
